group/assignment6/rearrange_cars.py

```diff
@@ -0,0 +1,103 @@
+    inverted_current_state correspondingly.
+
+    Returns:
+        new empty lot
+    """
+    #car = current_state[lot_from]
+    lot_from = inverted_current_state[car]
+    current_state[empty_lot] = car
+    current_state[lot_from] = 0
+    inverted_current_state[car] = empty_lot
+    inverted_current_state[0] = lot_from
+    return lot_from
+
+def place_car_to_lot(car, lot_to, current_state, inverted_state, empty_lot):
+    """Moves a given car from its lot to lot_to.
+
+    Returns:
+        new empty lot
+    """
+    #move_car_to_empty(lot_to, current_state, inverted_state, empty_lot)
+    empty_lot = move_car_to_empty(current_state[lot_to], current_state, inverted_state, empty_lot)
+    lot_from = inverted_state[car]
+    return move_car_to_empty(current_state[lot_from], current_state, inverted_state, empty_lot)
+
+def rearrange_cars_naive(start_state, end_state):
+    """Given a start state of n parking lots and n - 1 cars computes a sequence of moves that
+    rearranges the cars in order to obtain end state.
+    Number of moves in the worst case is 2 * n
+
+    Args:
+        start_state: a list of ints, each element corresponds to a car, zero corresponds to an
+        empty lot
+        end_state: same format as start_state
+
+    Returns:
+        moves_sequence: one possible sequence of moves
+    """
+    empty_lot = start_state.index(0)
+    current_state = start_state.copy()
+    inverted_state = [0 for i in range(len(start_state))]
+    moves_sequence = list()
+    for lot, car in enumerate(start_state):
+        inverted_state[car] = lot
+    for lot, car in enumerate(current_state):
+        # print(current_state)
+        if car == end_state[lot] or end_state[lot] == 0:
+            continue
+        new_car = end_state[lot]
+        # TODO: NamedTuple for move: (Car, From, To)
+        moves_sequence.append((car, lot, empty_lot))
+        moves_sequence.append((new_car, inverted_state[new_car], lot))
+        empty_lot = place_car_to_lot(new_car, lot, current_state, inverted_state, empty_lot)
+    # print(current_state)
+    # print(moves_sequence)
+    return moves_sequence
+
+
+def rearrange_cars(start_state, end_state):
+    """Given a start state of n parking lots and n - 1 cars computes a sequence of moves that
+    rearranges the cars in order to obtain end state where number of moves is minimal
+
+    Args:
+        start_state: a list of ints, each element corresponds to a car, zero corresponds to an
+        empty lot
+        end_state: same format as start_state
+
+    Returns:
+        moves_sequence: an optimal sequence of moves
+    """
+    empty_lot = start_state.index(0)
+    current_state = start_state.copy()
+    inverted_state = [0 for i in range(len(start_state))]
+    moves_sequence = list()
+    for lot, car in enumerate(start_state):
+        inverted_state[car] = lot
+    leftest_misplaced_car = 0
+    parking_size = len(start_state)
+
+    while leftest_misplaced_car < parking_size:
```

Comment:
Actually, I don't get the whole idea of this loop. I ran your script, and it turns out that this `while` loop executes only one step. In this step it goes to the nested `while` loop, where it performs all the moves, then in the conditional `if` below it executes all the additions in a row so that `leftest_misplaced_car` becomes equal to `parking_size`, then the third part is never executed because the condition does not hold anymore (`leftest_misplaced_car == parking_size`) and the whole outer `while` stops too.

I think you should rewrite it in a healthier way.

Reply:
Oh I see it now, you're right, thanks. It executes even 4 times, actually (depends on the position of empty slot)